scripts/visualize/visualize_utils.R

```diff
@@ -0,0 +1,160 @@
+
+visualize.svg_base_map <- function(viz){
+  geoms <- readDepends(viz)
+  library(svglite)
+  
+  # 1) set up shell svg, w/ proper size and aspect
+  # 2) add basic groups etc, including <defs><g id="template-geoms"/></defs> and <g id="styled-geoms"/>
+  # 3) set the plot bounds, including aspect of map
+  # 4) read in depends geoms
+  # 5) loop through depends, trim, then add geoms to <use/> elements (in id="template-geoms"), with id="u-{id}"
+  # 6) create geoms to mirror ids in <use/> elements, add attributes
+}
+
+#' calculate the bounding box of the sp object and return as `SpatialPolygons` object
+#' @param sp a spatial object
+#' 
+#' @return a `SpatialPolygons` object that represents the bounding box of the input `sp`
+get_sp_bbox <- function(sp){
+  bb <- bbox(sp)
+  xs <- bb[c(1, 3)]
+  ys <- bb[c(2, 4)]
+  proj.string = CRS(proj4string(sp))
+  return(as.sp_box(xs, ys, proj.string))
+}
+
+
+#' create a spatial polygon from x and y coordinates
+#' 
+#' @param xs a numeric vector of length two, containing the min and max of x values
+#' @param ys a numeric vector of length two, containing the min and max of y values
+#' 
+#' @return a `SpatialPolygons` object
+as.sp_box <- function(xs, ys, proj.string){
+  Sr1 <- Polygon(cbind(c(xs[c(1, 2, 2, 1, 1)]), c(ys[c(1, 1, 2, 2, 1)])))
+  Srs1 <- Polygons(list(Sr1), "s1")
+  SpP <- SpatialPolygons(list(Srs1), proj4string = proj.string)
+  return(SpP)
+}
+
+#' extract the svg elements from an sp object
+#' 
+#' @param sp a spatial object
+#' @param ... additional arguments passed to the plotting methods of `sp` (e.g., `expandBB`)
+#' @param width the width (in inches) of the svg view
+#' @param height the height (in inches) of the svg view
+#' @param pointsize number of pixels per inch
+#' @param xlim x limits (in sp units) of the plot
+#' @param ylim y limits (in sp units) of the plot
+#' 
+#' @return an `xml_document`
+get_svg_geoms <- function(sp, ..., width = 10, height = 8, pointsize = 12, xlim, ylim){
+  
+  stopifnot(packageVersion('svglite') == '1.2.0.9002')
+  
+  if (missing(xlim)){
+    xlim <- get_sp_lims(sp, ..., return = 'xlim')
+  }
+  if (missing(ylim)){
+    ylim <- get_sp_lims(sp, ..., return = 'ylim')
+  }
+  
+  # clip the spatial object so that it only contains features and data that are within the plotting range:
+  clipped.sp <- clip_sp(sp, xlim, ylim)
+  
+  rendered <- svglite::xmlSVG(width = width, height = height, pointsize = pointsize, standalone = F, {
+    set_sp_plot()
+    plot(clipped.sp, ..., xlim = xlim, ylim = ylim)
+  })
+  
+  sp.geoms <- tail(xml2::xml_children(rendered), length(sp))
```

Comment:
should this be `length(clipped.sp)`? or is there a way to filter to only those nodes that are type `path` instead, which might be a little clearer to non-svglite experts...or not...?

Reply:
good catch, yes, that should be `clipped.sp`

Reply:
As per comment on `path`, only the polygons render as paths. Lines will be either path or polyline (or line) elements, and points will be circles or paths depending on their `pch` value

Reply:
mm, ok, `clipped.sp` is sounding better than `path`